src/Citeproc/Eval.hs

```diff
@@ -1142,16 +1148,64 @@ evalItem layout (position, item) = do
            , contextPosition = position
            },
         st{ stateReference = ref
-          , stateUsedYearSuffix = False }))
+          , stateUsedYearSuffix = False
+          , stateUsedIdentifier = False
+          , stateUsedTitle = False
+          }))
         $ do xs <- mconcat <$> mapM eElement (layoutElements layout)
+
+             -- find identifiers that can be used to hyperlink the title 
+             let mbident = 
+                    foldl (<|>) Nothing
+                      [ IdentDOI   <$> (valToText =<< lookupVariable "DOI" ref)
+                      , IdentPMCID <$> (valToText =<< lookupVariable "PMCID" ref)
+                      , IdentPMID  <$> (valToText =<< lookupVariable "PMID" ref)
+                      , IdentURL   <$> (valToText =<< lookupVariable "URL" ref)
+                      ]
+             let mburl = identifierToURL <$> mbident
+            
+             -- hyperlink any titles in the output
+             let linkTitle url (Tagged TagTitle x) = Linked LinkTitle url [x]
+                 linkTitle _ x = x
+             
+            -- ensure correct handling of link prefixes like (https://doi.org/)
+            -- when a link's prefix+anchor=target, ensure the link includes the prefix
+            -- (see pandoc#6723 and citeproc#88; also pandoc's fixLinks function)
+             let fixLinkPrefixes (Formatted f [Linked lt u ys]) 
+                                 | checkPrefix f u ys = Linked lt u [Formatted f ys]
+                 fixLinkPrefixes y = y
+                 checkPrefix f url ys =
+                    let anchor = mconcat mempty (map outputToText ys)
+                        prefix = fromMaybe "" (formatPrefix f)
+                     in url == prefix <> anchor
+             
+             usedLink  <- gets stateUsedIdentifier
+             usedTitle <- gets stateUsedTitle
+             inBiblio  <- asks contextInBibliography 
+
+             -- when no links were rendered for a bibliography item, hyperlink
+             -- the title, if it exists, otherwise hyperlink the whole item
+             let xs' =
+                   if usedLink || not inBiblio
+                     then xs
+                   else case mburl of
+                         Nothing  -> xs
+                         Just url -> if usedTitle
+                                        -- hyperlink the title
+                                        then fmap (transform (linkTitle url)) xs
+                                        -- hyperlink the entire bib item
+                                        else [Linked LinkBibItem url xs] 
+            
+             let xs'' = fmap (transform fixLinkPrefixes) xs'
```

Comment:
Presumably we only need this transformation when `usedLink && inBiblio`?  So we could localize it more and avoid unnecessary work?

Or, maybe instead of having a traversal with `fixLinkPrefixes`, we could just build this logic into `renderOutput` (in Types)?

Reply:
Seems reasonable, so I made the change.